Fatras/include/ActsFatras/Kernel/SimulationResult.hpp

```diff
@@ -1,34 +1,63 @@
 // This file is part of the Acts project.
 //
-// Copyright (C) 2020 CERN for the benefit of the Acts project
+// Copyright (C) 2020-2021 CERN for the benefit of the Acts project
 //
 // This Source Code Form is subject to the terms of the Mozilla Public
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 #pragma once
 
+#include "ActsFatras/EventData/Hit.hpp"
+#include "ActsFatras/EventData/Particle.hpp"
+
+#include <iosfwd>
+#include <limits>
+#include <vector>
+
 namespace ActsFatras {
 
+/// Particle simulation status.
+enum class SimulationParticleStatus {
+  /// Particle is alive and can be simulated further.
+  eAlive,
+  /// Particle has been killed by an interaction.
+  eInteracted,
+  /// Particle has decayed.
+  eDecayed,
+};
+
+std::ostream& operator<<(std::ostream& os, SimulationParticleStatus status);
+
 /// Single particle simulation result (and intermediate state).
 ///
-/// This result struct is used by multiple components and thus defined
-/// separately.
+/// This result struct is used by multiple components and is thus defined here,
+/// separately from its usage.
 struct SimulationResult {
   /// Current/ final particle state.
   Particle particle;
-  /// Material accumulated during the propagation.
-  /// The initial particle can already have some accumulated material. The
-  /// particle stores the full material path. This keeps track of the
-  /// additional material accumulated during simulation.
-  Particle::Scalar pathInX0 = 0;
-  Particle::Scalar pathInL0 = 0;
-  /// Whether the particle is alive or not, i.e. could be simulated further.
-  bool isAlive = true;
+  /// Status flag, e.g. if the particle is alive and can be simulated further
+  SimulationParticleStatus particleStatus = SimulationParticleStatus::eAlive;
   /// Additional particles generated by interactions.
   std::vector<Particle> generatedParticles;
   /// Hits created by the particle.
   std::vector<Hit> hits;
+
+  // The following variables are internal simulation state implementation
+  // details that must be defined here for technical reasons.
+
+  // Accumulated particle proper time since the start of the simulation.
+  //
+  // Initialized to NaN so the simulation actor can detect when it is called
+  // for the first time, i.e. when the result struct is default-initialized.
+  Particle::Scalar properTime =
+      std::numeric_limits<Particle::Scalar>::quiet_NaN();
+  // Accumulated radiation/interaction length limit before next interaction.
+  Particle::Scalar x0Limit = std::numeric_limits<Particle::Scalar>::infinity();
+  Particle::Scalar l0Limit = std::numeric_limits<Particle::Scalar>::infinity();
+  // Process selection for the next interaction.
+  size_t x0Process = SIZE_MAX;
+  size_t l0Process = SIZE_MAX;
```

Comment:
The X0/L0 limits can not be members of the `Interactor` since they can be reset during the simulation and the `Interactor` must be a `const` object; if a point-like interaction does not kill the particle, the limits need to be updated to simulate the next interaction and so on. This is not the case for the decay limit. This can only happen once and can thus be fixed before the simulation starts (and the limit can be a member of the interactor).

Reply:
Right, it could be modified. But it is way easier to read and understand if one doesn't have to handle several files in order to get all possible limits. Therefore I would argue to keep the time limit inside the result object and initialise it alongside the others. Then you have logical blocks of code like "here are limits stored", "here we set limits", "here we handle a reached limit"

Reply:
In the most recent versions, all limits are stored in the same struct and are handled by the `Interactor`.